README.md
gHSS
=====

This software implements algorithms for the incremental greedy approximation to the Hypervolume Subset Selection Problem (HSSP) in two and three dimensions (gHSS2D and gHSS3D [1], respectively), which provide an approximation guarantee of 1-1/e. Minimization is assumed.


**Note**: Although only *nondominated* points that strongly dominate the reference point contribute to the value of the hypervolume indicator, the code is prepared to deal with all other points, including *repeated* points. Warnings will be raised if any of the points do not strongly dominate the reference point. Moreover, different points may have (some) equal coordinates.


**Note 2**: In two dimensions, the exact solution can also be computed efficiently, see [here](https://eden.dei.uc.pt/~paquete/HSSP/) and [here](http://hpi.de/friedrich/docs/code/ssp.zip).

License
--------


Except where indicated otherwise in individual source files, this software is Copyright © 2016, 2017 Andreia P. Guerreiro.

This program is free software. You can redistribute it and/or modify it under the terms of the GNU General Public License, version 3, as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

Appropriate reference to this software should be made when describing research in which it played a substantive role, so that it may be replicated and verified by others. The algorithms which this software implements are described in detail in [1]. 





Building
--------


In GNU/Linux, the program can be compiled from source by invoking:

    make

It is recommended that you compile it specifically for your architecture. Depending on the compiler and version of the compiler you use, there are different ways to achieve this. For recent GCC versions, make will pick a suitable -march argument based on the processor of the build machine. This can be overridden by passing a `MARCH=` argument to make. Similarly if you use the Intel C compiler, it will pick a sensible default architecture (`-xHOST`) for you. If you want to override this, pass `XARCH=` to `make`. So, to build for an Intel *Core2* with the GCC compiler, you would use:

    make MARCH=core2

For the Intel C compiler, you should use:

    make XARCH=SSSE3

Generally, `make` will try to pick good flags for you, but you can override them by passing an `OPT_CFLAGS` argument to `make` if you wish. To build an unoptimized version of `hv4d` you could run:

    make OPT_CFLAGS="-O0 -g"

Finally, if you do not want to see the command line of each compiler invocation, pass `S=1` to make.



General Usage
-------------


**SYNOPSIS** 

    gHSS [OPTIONS] [FILE...]
    
        
**DESCRIPTION**

Compute the incremental greedy approximation to the hypervolume subset selection problem for the data set(s) in FILE(s).

With no FILE, read from the standard input.

**COMMAND LINE OPTIONS**

	 -h, --help          print this summary and exit.                          
	     --version       print version number and exit.                        
	 -v, --verbose       print some information (time, coordinate-wise maximum and minimum, etc)                                     
	 -q, --quiet         print only the results (as opposed to --verbose). 
	 -u, --union         treat all input sets within a FILE as a single set.   
	 -r, --reference=POINT use POINT as the reference point. POINT must be within quotes, e.g.,
		                 "10 10 10". If no reference point is given, it is taken as the
		                 coordinate-wise maximum of all input points.                                     
	 -s, --suffix=STRING Create an output file for each input file by appending this suffix.
		                 This is ignored when reading from stdin. If missing, output is sent
		                 to stdout.             
	 -k, --subsetsize=k  select k points (a value between 1 and n, where n is the size of the
		                 input data set. The default is n/2)   
	 -f, --format=(0|..|4) output format
		                 (0: print indices followed by the hypervolume indicator of the selected subset (default))        
		                 (1: print indices of the selected points)             
		                 (2: print the hypervolume indicator of the selected subset)    
		                 (3: print indices and the corresponding contributions to the previous subset)
		                 (4: print indices and the corresponding accumulated hypervolume)           
		                        
                               

       

Usage
-----

**Run**

The program reads sets of points from the file(s) specified in the command line:

    ./gHSS data

or standard input:

    cat data | ./gHSS

In input files, each point is given in a separate line, and point coordinates in each line are separated by whitespace. An empty line, or a line beginning with a  hash sign (#), denotes a separate set.


Sets in an input file may be treated as a single set by using option `-u`: 

    ./gHSS data -u


The reference point can be set by giving option `-r`.

    ./gHSS -r "10 10 10 10" data

 If no reference point is given, the default is the coordinate-wise maximum of all input points in all files.

For the other options available, check the output of `./gHSS --help`.
 
By default, the default subset size *k* is set to *n/2* where *n* is the data set size. The subset size can be explicitly specified using option `-k`:
 
    ./gHSS -r "10 10 10 10" data -k 10

For the other options available, check the output of `./gHSS --help`.



Examples
-------


**Input File(s)**

Empty lines and lines starting with a hash sign (#) at the beginning and/or at the end of the file are ignored.

Example of valid content of input files:

    1   1   4   4
    4   4   1   1
    
    2   2   3   3
    3   3   2   2

Another example:

    #
    6 4 9 6
    7 3 7 1
    8 2 3 2
    9 1 2 4
    #
    1 9 5 4
    2 8 3 1
    #
    3 7 8 1
    4 6 3 9
    5 5 5 8
    #
    
       
**Compilation**

Example of basic compilation: 

    make march=corei7

    
**Execution**

The code can be tested with any of the data sets provided in folder `examples`. File `test.inp` in that folder contains a 3-dimensional example with the following set of *10* points:

    0.16 0.86 0.47 
    0.66 0.37 0.29 
    0.79 0.79 0.04 
    0.28 0.99 0.29 
    0.51 0.37 0.38
    0.92 0.62 0.07 
    0.16 0.53 0.70 
    0.01 0.98 0.94 
    0.67 0.17 0.54 
    0.79 0.72 0.05 



After compilation, to compute the greedy solution for a subset of size 5 given the reference point `(1, 1, 1)`, run:

    ./gHSS examples/test.inp -r "1 1 1" -k 5

which produces the following output:

    4
    6
    8
    9
    1
    0.304494   


In the *default* output format (`-f 0`), the first *k* lines contain the indices (between *0* and *n-1*) of the selected points in the order in which they were selected. In the above example, the first point selected was the (4+1)-th point in the file, i.e., `(0.51, 0.37, 0.38)`, the second point selected was the (6+1)-th point in the file, i.e., `(0.16, 0.53, 0.70)`, and so forth. The last line contains the hypervolume indicator of the greedy solution, i.e., of the subset containing the *k* selected points. If option `-f 1` is given, the outptut contains only the indices (i.e, the first *k* lines in option `-f 0`). If option `-f 2` is given instead, only the hypervolume indicator of the selected subset is printed.

If option `-f 3`is given:


    ./gHSS examples/test.inp -r "1 1 1" -k 5 -f 3
    
the corresponding output is:


    4       0.191394       
    6       0.04935        
    8       0.03036        
    9       0.019404       
    1       0.013986



The output contains *k* lines with two columns. The first column is the same as for option `-f 1`. The second column shows the contribution of the corresponding point to the subset containing the points previously selected. In the example above, the first point selected was point `(0.51, 0.37, 0.38)`, and its contribution (to the empty set) was `0.191394`. The last point selected (i.e., point `(0.66, 0.37, 0.29)`) contributed `0.013986` to the hypervolume indicator of the previous subset of *k-1* points. Note that, since each point selected contributes the most to the previous subset, the second column is always in nonincreasing order.

Finally, if option `-f 4` is given, i.e.,

    ./gHSS examples/test.inp -r "1 1 1" -k 5 -f 4

The corresponding output is:

    4       0.191394       
    6       0.240744       
    8       0.271104       
    9       0.290508       
    1       0.304494   

In this case, the second column shows the accumulated hypervolume. Consequently, the second column of the *i*-th line shows the hypervolume indicator of the subset of the first *i* points selected.



References
----------


[1] A. P. Guerreiro, C. M. Fonseca, and L. Paquete, “Greedy hypervolume subset selection in low dimensions,” Evolutionary Computation, vol. 24, pp. 521-544, Fall 2016. [ [DOI](http://dx.doi.org/10.1162/EVCO_a_00188) | [pdf](https://eden.dei.uc.pt/~cmfonsec/GreedyHSS-ECJ2016-authorVersion.pdf) ] 

[2] A. P. Guerreiro, C. M. Fonseca, and L. Paquete, “Greedy hypervolume subset selection in the three-objective case,” in Proceedings of the 2015 on Genetic and Evolutionary Computation Conference, GECCO '15, (Madrid, Spain), pp. 671-678, ACM, 2015. EMO Track Best Paper Award. [ [DOI](http://dx.doi.org/10.1145/2739480.2754812) ]



 





CHANGELOG.md
#Change Log

<!-- ## [Unreleased] -->

## [1.1] - 2017-02-08
#### Fixed
- Points that do not dominate the reference point could lead to incorrect results.
- Bug related to equal coordinates in two-dimensional problems.

#### Changed
- Raise a warning instead of an error if no point strongly dominates the reference point.

#### Added
- More output format options.

## 1.0 - 2015
- Original version.

[1.1]: https://github.com/apguerreiro/gHSS/releases/tag/v1.1


io.h
#ifndef _GHSS_IO_H_
#define _GHSS_IO_H_

#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>

#define point_printf_format "%-16.15g"

/* If we're not using GNU C, elide __attribute__ */
#ifndef __GNUC__
#  define  __attribute__(x)  /* NOTHING */
#endif

void 
errprintf(const char * template,...) 
/* enables the compiler to check the format string against the
   parameters */  __attribute__ ((format(printf, 1, 2)));

void warnprintf(const char *template,...)
/* enables the compiler to check the format string against the
   parameters */  __attribute__ ((format(printf, 1, 2)));

/* Error codes for read_data.  */
#define READ_INPUT_FILE_EMPTY -1
#define READ_INPUT_WRONG_INITIAL_DIM -2

int
read_data (const char *filename, double **data_p, 
           int *nobjs_p, int **cumsizes_p, int *nsets_p);

#endif


gHSS.h
/*************************************************************************

 ghss.h

 ---------------------------------------------------------------------

                       Copyright (c) 2016, 2017
                Andreia P. Guerreiro <apg@dei.uc.pt>
             

 This program is free software (software libre); you can redistribute
 it and/or modify it under the terms of the GNU General Public License,
 version 3, as published by the Free Software Foundation.

 This program is distributed in the hope that it will be useful, but
 WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, you can obtain a copy of the GNU
 General Public License at:
                 http://www.gnu.org/copyleft/gpl.html
 or by writing to:
           Free Software Foundation, Inc., 59 Temple Place,
                 Suite 330, Boston, MA 02111-1307 USA



*************************************************************************/
#ifndef GHSS_H_
#define GHSS_H_

#ifdef __cplusplus
extern "C" {
#endif

double greedyhss(double *data, int d, int n, const int k, const double *ref, double * volumes, int * selected);

#ifdef __cplusplus
}
#endif

#endif


main-gHSS.c
/*************************************************************************

 gHSS: main program

 ---------------------------------------------------------------------

                        Copyright (c) 2016, 2017
                Andreia P. Guerreiro <apg@dei.uc.pt>
             
                       Copyright (c) 2010
                Carlos M. Fonseca <cmfonsec@dei.uc.pt>
           Manuel Lopez-Ibanez <manuel.lopez-ibanez@ulb.ac.be>
                  Luis Paquete <paquete@dei.uc.pt>
                Andreia P. Guerreiro <apg@dei.uc.pt>

 This program is free software (software libre); you can redistribute
 it and/or modify it under the terms of the GNU General Public License,
 version 3, as published by the Free Software Foundation.

 This program is distributed in the hope that it will be useful, but
 WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, you can obtain a copy of the GNU
 General Public License at:
                 http://www.gnu.org/copyleft/gpl.html
 or by writing to:
           Free Software Foundation, Inc., 59 Temple Place,
                 Suite 330, Boston, MA 02111-1307 USA

 ----------------------------------------------------------------------

 Relevant literature:

 [1] A. P. Guerreiro, C. M. Fonseca, and L. Paquete, “Greedy hypervolume subset selection in low dimensions,” Evolutionary Computation, vol. 24, Fall 2016. In press.
 [2] A. P. Guerreiro, C. M. Fonseca, and L. Paquete, “Greedy hypervolume subset selection in the three-objective case,” in Proceedings of the 2015 on Genetic and Evolutionary Computation Conference, GECCO '15, (Madrid, Spain), pp. 671--678, ACM, 2015.

*************************************************************************/
#include "io.h"
#include "gHSS.h"
#include "timer.h"

#include <errno.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <ctype.h> // for isspace()

#include <unistd.h>  // for getopt()
#include <getopt.h> // for getopt_long()
#include <limits.h>

#ifdef __USE_GNU
extern char *program_invocation_short_name;
#else
char *program_invocation_short_name;
#endif

static const char * const stdin_name = "<stdin>";
static int verbose_flag = 1;
static bool union_flag = false;
static char *suffix = NULL;
static int ksub = -1;
static int outflag = 0; //0 - index and hypervolume (default), 1 - index, 2 - hypervolume, 3 - index and contribution, 4 - accumulated contribution


static void usage(void)
{
    printf("\n"
           "Usage: %s [OPTIONS] [FILE...]\n\n", program_invocation_short_name);

    printf(
"Calculate the hypervolume of each input set of each FILE. \n"
"With no FILE, or when FILE is -, read standard input.\n\n"

"Options:\n"
" -h, --help           print this summary and exit.                          \n"
"     --version        print version number and exit.                        \n"
" -v, --verbose        print some information (time, coordinate-wise maximum \n"
"                      and minimum, etc)                                     \n"
" -q, --quiet          print only the results (as opposed to --verbose).     \n"
" -u, --union          treat all input sets within a FILE as a single set.   \n"
" -r, --reference=POINT use POINT as the reference point. POINT must be within\n"
"                      quotes, e.g., \"10 10 10\". If no reference point is  \n"
"                      given, it is taken as the coordinate-wise maximum of  \n"
"                      all input points.                                     \n"
/*
"                      given, it is taken as max + 0.1 * (max - min) for each\n"
"                      coordinate from the union of all input points.        \n"
*/
" -s, --suffix=STRING  Create an output file for each input file by appending\n"
"                      this suffix. This is ignored when reading from stdin. \n"
"                      If missing, output is sent to stdout.                 \n"
" -k, --subsetsize=k   select k points (a value between 1 and n, where n is  \n"
"                      the size of the input data set. The default is n/2)   \n"                     
" -f, --format=(0|..|4) output format                                        \n"
"                      (0: print indices followed by the hypervolume         \n"
"                         indicator of the selected subset (default))        \n"
"                      (1: print indices of the selected points)             \n"
"                      (2: print the hypervolume indicator of the selected   \n"
"                         subset)                                            \n"
"                      (3: print indices and the corresponding contributions \n"
"                         to the previous subset)                            \n"
"                      (4: print indices and the corresponding accumulated   \n"
"                         hypervolume)                                       \n"
"\n");

}

static void version(void)
{
    printf("%s version " VERSION
#ifdef ARCH
           " (optimised for " ARCH ")"
#endif
           "\n\n", program_invocation_short_name);

    printf(
"Copyright (c) 2016, 2017"
"\nAndreia P. Guerreiro <apg@dei.uc.pt>\n"
"\n"
"This is free software, and you are welcome to redistribute it under certain\n"
"conditions.  See the GNU General Public License for details. There is NO   \n"
"warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"
"\n"        );
}

static inline void
vector_printf (const double *vector, int size)
{
    int k;
    for (k = 0; k < size; k++)
        printf (" %f", vector[k]);
}

static double *
read_reference(char * str, int *nobj)
{
    double * reference;
    char * endp;
    char * cursor;

    int k = 0, size = 10;

    reference = malloc(size * sizeof(double));
    endp = str;

    do {
        cursor = endp;
        if (k == size) {
            size += 10;
            reference = realloc(reference, size * sizeof(double));
        }
        reference[k] = strtod(cursor, &endp);
        k++;
    } while (cursor != endp);

    // not end of string: error
    while (*cursor != '\0') {
        if (!isspace(*cursor)) return NULL;
        cursor++;
    }

    // no number: error
    if (k == 1) return NULL;

    *nobj = k-1;
    return reference;
}

static inline void
handle_read_data_error (int err, const char *filename)
{
    switch (err) {
    case 0: /* No error */
        break;

    case READ_INPUT_FILE_EMPTY:
        errprintf ("%s: no input data.", filename);
        exit (EXIT_FAILURE);

    case READ_INPUT_WRONG_INITIAL_DIM:
        errprintf ("check the argument of -r, --reference.\n");
    default:
        exit (EXIT_FAILURE);
    }
}

static void
data_range (double **maximum, double **minimum,
            const double *data, int nobj, int rows)
{
    int n, k = 0;
    int r = 0;

    if (*maximum == NULL) {
        *maximum = malloc (nobj*sizeof(double));
        for (k = 0; k < nobj; k++)
            (*maximum)[k] = data[k];
        r = 1;
    }

    if (*minimum == NULL) {
        *minimum = malloc (nobj*sizeof(double));
        for (k = 0; k < nobj; k++)
            (*minimum)[k] = data[k];
        r = 1;
    }

    for (; r < rows; r++) {
        for (n = 0; n < nobj; n++, k++) {
            if ((*maximum)[n] < data[k])
                (*maximum)[n] = data[k];
            if ((*minimum)[n] > data[k])
                (*minimum)[n] = data[k];
        }
    }
}

static void
file_range (const char *filename, double **maximum_p, double **minimum_p,
            int *dim_p)
{
    double *data = NULL;
    int *cumsizes = NULL;
    int nruns = 0;
    int dim = *dim_p;
    double *maximum = *maximum_p;
    double *minimum = *minimum_p;

    handle_read_data_error (
        read_data (filename, &data, &dim, &cumsizes, &nruns), filename);

    data_range (&maximum, &minimum, data, dim, cumsizes[nruns-1]);

    *dim_p = dim;
    *maximum_p = maximum;
    *minimum_p = minimum;

    free (data);
    free (cumsizes);
}

/*
   FILENAME: input filename. If NULL, read stdin.

   REFERENCE: reference point. If NULL, use MAXIMUM.

   MAXIMUM: maximum objective vector. If NULL, caculate it from the
   input file.

   NOBJ_P: pointer to number of objectives. If NULL, calculate it from
   input file.

*/

static void
gHSS_file (const char *filename, double *reference,
         double *maximum, double *minimum, int *nobj_p)
{
    double *data = NULL;
    int *cumsizes = NULL;
    int cumsize;
    int nruns = 0;
    int n, k;
    int nobj = *nobj_p;
    char *outfilename = NULL;
    FILE *outfile = stdout;
    bool setmax = false;
    bool setref = false;

    int err = read_data (filename, &data, &nobj, &cumsizes, &nruns);
    if (!filename) filename = stdin_name;
    handle_read_data_error (err, filename);

    if (filename != stdin_name && suffix) {
        int outfilename_len = strlen(filename) + strlen(suffix) + 1;

        outfilename = malloc (sizeof(char) * outfilename_len);
        strcpy (outfilename, filename);
        strcat (outfilename, suffix);

        outfile = fopen (outfilename, "w");
        if (outfile == NULL) {
            errprintf ("%s: %s\n", outfilename, strerror(errno));
            exit (EXIT_FAILURE);
        }
    }

    if (union_flag) {
        cumsizes[0] = cumsizes[nruns - 1];
        nruns = 1;
    }

    if (verbose_flag == 2)
        printf("# file: %s\n", filename);

    if (maximum == NULL) {
        setmax = true;
        data_range (&maximum, &minimum, data, nobj, cumsizes[nruns-1]);
        if (verbose_flag == 2) {
            printf ("# maximum:");
            vector_printf (maximum, nobj);
            printf ("\n");
            printf ("# minimum:");
            vector_printf (minimum, nobj);
            printf ("\n");
        }
    }

    if (reference != NULL) {
        for (n = 0; n < nobj; n++) {
            if (reference[n] <= maximum[n]) {
                warnprintf ("%s: some points do not strictly dominate "
                            "the reference point",
                            filename);
                break;
            }
        }
    } else {
        setref = true;
        reference = malloc(nobj * sizeof(double));
        for (n = 0; n < nobj; n++) {
            /* default reference point is: */
            reference[n] = maximum[n];
            /* however, a better reference point is:
            reference[n] = maximum[n] + 0.1 * (maximum[n] - minimum[n]);
            so that extreme points have some influence. */
        }
    }

    if (verbose_flag == 2) {
        printf ("# reference:");
        vector_printf (reference, nobj);
        printf ("\n");
    }

    for (n = 0, cumsize = 0; n < nruns; cumsize = cumsizes[n], n++) {
        double time_elapsed_cpu;
        double volume;
        int i;

        if (verbose_flag == 2)
            fprintf (outfile, "# Data set %d:\n", n + 1);

        int size = cumsizes[n] - cumsize;
        
           
        k = (ksub > 0) ? ((ksub < size) ? ksub : size) : size/2;
        double * volumes = (double *) malloc(k * sizeof(double));
        int * selected = (int *) malloc(k * sizeof(int));
        
        
        Timer_start ();
        volume = greedyhss(&data[nobj * cumsize], nobj, size, k,
                                    reference, volumes, selected);
        
        time_elapsed_cpu = Timer_elapsed_virtual ();
        
        
        if (volume <= 0.0) {
//             errprintf ("none of the points strictly dominates the reference point\n");
//             exit (EXIT_FAILURE);
            warnprintf ("none of the points dominates the reference point\n");
        }

        
        if (verbose_flag == 2 && outflag != 2) fprintf (outfile, "# index");
        
        double volk = 0;
        switch(outflag){
            
            case 0: //print indices as in case 1
            case 1:
            case 2:
                
                if (verbose_flag == 2 && outflag != 2) fprintf (outfile, "\n");
                for(i = 0; i < k; i++){
                    volk += volumes[i];
                    if(outflag != 2) fprintf (outfile, "%d\n", selected[i]);
                }
                if(outflag != 1){
                    if(verbose_flag == 2) fprintf (outfile, "# hypervolume indicator\n");
                    fprintf(outfile, "%-16.15g\n", volk);
                }
                break;
            
            case 3:
                
                if (verbose_flag == 2) fprintf (outfile, " contribution\n");
                for(i = 0; i < k; i++){
                    fprintf (outfile, "%d\t%-16.15g\n", selected[i], volumes[i]);
                }
                break;
                
            case 4:
                if (verbose_flag == 2) fprintf (outfile, " accumulated_contribution\n");
                for(i = 0; i < k; i++){
                    volk += volumes[i];
                    fprintf(outfile, "%d\t%-16.15g\n", selected[i], volk);
                }
                break;
                
            default:
                fprintf(outfile, "wrong flag code! %d\n", outflag);
        }
        
    
        
        free(volumes);
        free(selected);



        if (verbose_flag == 2) {
            fprintf (outfile, "# Time computing gHSS (cpu): %f seconds\n", time_elapsed_cpu);
        }else if(verbose_flag == 3) {
            fprintf (outfile, "%f\n", time_elapsed_cpu);
            
//             fprintf (outfile, "-> %f\n", time_elapsed_cpu);
        }
        if(nruns > 1) fprintf(outfile, "\n");
        

    }

    if (outfilename) {
        if (verbose_flag)
            fprintf (stderr, "# %s -> %s\n", filename, outfilename);
        fclose (outfile);
        free (outfilename);
    }
    free (data);
    free (cumsizes);
    if (setmax){
        free (maximum);
        free (minimum);
    }
    if (setref) free(reference);
    *nobj_p = nobj;
}

int main(int argc, char *argv[])
{
    double *reference = NULL;
    double *archiveParam = NULL;
    int nobj = 0;
    int numfiles, k;

    int opt; /* it's actually going to hold a char.  */
    int longopt_index;

    /* see the man page for getopt_long for an explanation of these fields.  */
    static struct option long_options[] = {
        {"help",       no_argument,       NULL, 'h'},
        {"version",    no_argument,       NULL, 'V'},
        {"verbose",    no_argument,       NULL, 'v'},
        {"quiet",      no_argument,       NULL, 'q'},
        {"reference",  required_argument, NULL, 'r'},
        {"union",      no_argument,       NULL, 'u'},
        {"suffix",     required_argument, NULL, 's'},
        {"subsetsize", required_argument, NULL, 'k'},
        {"format",     required_argument, NULL, 'f'},

        {NULL, 0, NULL, 0} /* marks end of list */
    };

#ifndef __USE_GNU
    program_invocation_short_name = argv[0];
#endif

    while (0 < (opt = getopt_long (argc, argv, "hVvqur:s:k:f:",
                                   long_options, &longopt_index))) {
        switch (opt) {
        case 'r': // --reference
            reference = read_reference (optarg, &nobj);
            if (reference == NULL) {
                errprintf ("invalid reference point '%s",
                           optarg);
                exit (EXIT_FAILURE);
            }
            break;

        case 'u': // --union
            union_flag = true;
            break;

        case 's': // --suffix
            suffix = optarg;
            break;

        case 'V': // --version
            version();
            exit(EXIT_SUCCESS);

        case 'q': // --quiet
            if(verbose_flag == 1) //no option -v
                verbose_flag = 0;
            else  //option -q and -v
                verbose_flag = 3;
            break;

        case 'v': // --verbose
            if(verbose_flag == 1) //no option -q
                verbose_flag = 2;
            else  //option -q and -v
                verbose_flag = 3;
            break;

        case 'k':
            ksub = (strcmp(optarg, "MAX") == 0) ? INT_MAX : atoi(optarg);
            break;
            
        case 'f': // 
            // outflag: 0 - index of selected points and the hypervolume indicator of that subset (default)
            // outflag: 1 - index of selected points
            // outflag: 2 - the hypervolume indicator of that subset
            // outflag: 3 - index of selected points and corresponding contribuition at the time the point was selected
            // outflag: 4 - index of selected points and accumulated contribution
            outflag = atoi(optarg);
            break;
            
            
        case '?':
            // getopt prints an error message right here
            fprintf (stderr, "Try `%s --help' for more information.\n",
                     program_invocation_short_name);
            exit(EXIT_FAILURE);
        case 'h':
            usage();
            exit(EXIT_SUCCESS);

        default: // should never happen
            abort();
        }
    }

    numfiles = argc - optind;

    if (numfiles < 1) /* Read stdin.  */
        gHSS_file (NULL, reference, NULL, NULL, &nobj);

    else if (numfiles == 1) {
        gHSS_file (argv[optind], reference, NULL, NULL, &nobj);
    }
    else {
        double *maximum = NULL;
        double *minimum = NULL;
        if (reference == NULL) {
            /* Calculate the maximum among all input files to use as
               reference point.  */
            for (k = 0; k < numfiles; k++)
                file_range (argv[optind + k], &maximum, &minimum, &nobj);

            if (verbose_flag == 2) {
                printf ("# maximum:");
                vector_printf (maximum, nobj);
                printf ("\n");
                printf ("# minimum:");
                vector_printf (minimum, nobj);
                printf ("\n");
            }
        }
        for (k = 0; k < numfiles; k++)
            gHSS_file (argv[optind + k], reference, maximum, minimum, &nobj);
        
        free (maximum);
        free (minimum);
    }
    free(archiveParam);
    
    if (reference != NULL) free (reference);
    return EXIT_SUCCESS;
}
